### DISCUSSION TITLE
Proposal for a Consistent Repository Naming Method for Template Instantiation Solutions

### DISCUSSION THREAD

**Discussion by msclock** [serious-scaffold/.github#1] [Ideas]
Recently, we have been discussing the need for a consistent agreement on the repository naming method for our template instantiation solutions. In order to ensure efficient management and organization of our repositories in the future, I propose that we establish a standard naming convention. 
 
The goal of this proposal is to streamline our repository processes and enhance clarity and accessibility for all team members. By adhering to a consistent naming method, we can avoid confusion, improve searchability, and facilitate seamless collaboration. 
 
I would like to initiate a discussion on this matter to gather your valuable insights and suggestions. Here are a few points to consider: 
 
1. What are the current challenges we face in managing our template instantiation repositories? 
2. How can a standardized naming convention address these challenges and improve efficiency? 
3. Are there any existing naming conventions or best practices in the industry that we can draw inspiration from? 
4. What factors should we consider when designing a naming method? (e.g., project-specific details, versioning, categorization) 
5. How can we ensure the proposed naming convention is easily scalable and adaptable for future needs? 
 
I encourage everyone to share their thoughts, experiences, and any relevant examples. Let's work together towards a consensus on the repository naming method that will benefit our team and the overall project. 
 
Thank you for your time and contributions.

**Comment by huxuan** [comment]
Really appreciate the initiation of this discussion.

Currently we have two main repos: `serious-scaffold-python` and `cpp-scaffold`.

IMO, `serious-scaffold-python` is a little bit long especially when combining with the organization name, for example, when generating a project, the command `copier copy gh:serious-scaffold/serious-scaffold-python /path/to/project` is a little bit long. For `cpp-scaffold`, the problem is a little bit generic, unless used together with the organization name.

As we have discussed, we will have bunch of templates by leveraging the [multiple templates](https://copier.readthedocs.io/en/stable/configuring/#applying-multiple-templates-to-the-same-subproject) mechanism. 

For Python-only series, what I can imagine is at least three templates:
1. Python Scaffold
1. FastAPI Scaffold
1. PyTorch Scaffold

For Python Binding series, there are at least two categories:
1. C++ Python Binding, but there could be many approaches, for example, `Boost.Python`, `pybind11`.
1. Rust Python Bindings, seems `pyo3` is the only recommended choice?

Considering our requirements, seems there are two points we need to make decision:

1. Do we want to cover languages other than Python related ecosystem? For example, cpp-only projects, Java projects and even some frontend languages like nodejs projects.
1. Do we want to keep the `serious` term or something else? Before coming up with the `serious` idea, I know there is a popular Python Scaffold project naming [pyscaffold](https://github.com/pyscaffold/pyscaffold).

BTW, we also need to check that the package name is available on [PyPI](https://pypi.org).

**Reply by msclock** [reply]
@huxuan  Indeed, there are  many industry solutions based on languages-specific. Considering language-specific series, perhaps it would be easily scalable and adaptable to standardize naming convention using language abbr as prefixes and language-specific solutions as suffixes. For example:

In python domain, use `py-<language-specific-solution>`: 
- `py-base`: base scaffold for developing simple package and cmd tools.(not very sure about that and maybe it is more reasonable to use py-package or py-scaffold)
- `py-fastapi`: fastapi scaffold for developing fastapi server.
- `py-torch`: pytorch scaffold for developing torch modules.
- `py-boost-python`: cpp binding developed based on Boost.Python.
- `py-pybind11`: cpp binding developed based on pybind11.
- ...
In cpp domain, use `cpp-<language-secific-solution>`: 
- `cpp-base`: base scaffold for developing simple package and cmd tools.(or cpp-package, cpp-scaffold)
- `cpp-jni`: cpp binding based on java jni.
- `cpp-pybind11`: cpp binding based on pybind11.🤣
- `cpp-swig`: cpp binding based on swig.
- ...

This is just my initial idea about naming convention. If you have any more suggestions or considerations, please feel free to bring them up for discussion.😎

**Reply by huxuan** [reply]
I would suggest to have some word or phrase to differentiate our templates from others and only using `scaffold` seems not to be enough. For example, `py-torch` is kind of ambiguous to the real `pytorch` and `pyscaffold` is already occupied. Maybe we should have something like `pyo3`? There is a reddit thread about this: https://www.reddit.com/r/rust/comments/ddurtp/how_pyo3_got_that_name/.

**Reply by msclock** [reply]
How about using ss(an abbr of Serious Scaffold) as prefixes, for example. sspy or pyss. I think sspy is more  reasonable and more expressive.

**Reply by huxuan** [reply]
That looks good to me. To double confirm, so it will be

* `sspy`
* `sspy-fastapi`
* `sspy-torch`
* `sspy-boost`
* `sspy-pybind11`
* `sscpp`
* `sscpp-jni`
* `sscpp-pybind11`
* `sscpp-swig`
* `sspyo3`

right? The only problem for me is that there are two `pybind11`, any difference between them? Or in other words, if the scaffold is based on two languages, cpp & Python or Rust & Python, which one should we use?

Also I notice that both [`sspy`](https://pypi.org/project/sspy/) and [`ss-py`](https://pypi.org/project/ss-py/) are already occupied. I will take a look whether we can take it over.

**Reply by msclock** [reply]
> if the scaffold is based on two languages, cpp & Python or Rust & Python, which one should we use?

Perhaps it could transfer to `ss<language-specific>` series. After all, the development languages and tools ​​​​that each is good at vary from person to person and ecosystem to ecosystem.

**Reply by huxuan** [reply]
Then what is the difference between `sspy-pubind11` and `sscpp-pybind11`? Or maybe we can just remove the language name for the combination cases, for example, `ss-pybind11`? We can use use the same pattern for framework specific templates, for examlpe, `ss-fastapi` and `ss-pytorch`. Any comments?

**Reply by msclock** [reply]
This is a good point, given that a specific framework is more reasonable and maintainable, we can provide a quick implementation and develop it to see if it is what we expect.

**Reply by huxuan** [reply]
Great, then at least we can reach an agreement on the 

- ss-python
- ss-cpp

Is that right?

For those combination and frame specific cases, we can just talk by code.

**Reply by msclock** [reply]
Great, let's move forward.

**Reply by huxuan** [reply]
BTW, I am trying to ask whether we can have the package name [`ss-py`](https://pypi.org/project/ss-py/), If that works, the name will be more consistent as both `py` and `cpp` are abbrevations.

**Reply by msclock** [reply]
Actually, I don't think it is necessary to upload template to official pypi(https://pypi.org/). But there is a test one(https://test.pypi.org/) that is less used by community.🤣 And I can confirm there is no ss-py(https://test.pypi.org/search/?q=ss-py).

**Reply by huxuan** [reply]
> Actually, I don't think it is necessary to upload template to official pypi(https://pypi.org/).

That is a great point! We can even setup a new one for ourselves since there is no distribution demand.
The only problem I can come up with is that some badges may not work then, but that makes our scaffold more robust.
